xdggs/tests/test_h3.py

```diff
@@ -0,0 +1,132 @@
+resolutions = [1, 5, 15]
+variable_names = ["cell_ids", "zonal_ids", "zone_ids"]
+
+variables = [
+    xr.Variable(
+        dims[0], cell_ids[0], {"grid_type": "h3", "resolution": resolutions[0]}
+    ),
+    xr.Variable(
+        dims[1], cell_ids[0], {"grid_type": "h3", "resolution": resolutions[0]}
+    ),
+    xr.Variable(
+        dims[0], cell_ids[1], {"grid_type": "h3", "resolution": resolutions[1]}
+    ),
+    xr.Variable(
+        dims[1], cell_ids[2], {"grid_type": "h3", "resolution": resolutions[2]}
+    ),
+]
+variable_combinations = [
+    (old, new) for old, new in itertools.product(variables, repeat=2)
+]
+
+
+@pytest.mark.parametrize("resolution", resolutions)
+@pytest.mark.parametrize("dim", dims)
+@pytest.mark.parametrize("cell_ids", cell_ids)
+def test_init(cell_ids, dim, resolution):
+    index = h3.H3Index(cell_ids, dim, resolution)
+
+    assert index._resolution == resolution
+    assert index._dim == dim
+
+    # TODO: how do we check the index, if at all?
+    assert index._pd_index.dim == dim
+    assert np.all(index._pd_index.index.values == cell_ids)
+
+
+@pytest.mark.parametrize("variable", variables)
+@pytest.mark.parametrize("variable_name", variable_names)
+@pytest.mark.parametrize("options", [{}])
+def test_from_variables(variable_name, variable, options):
+    expected_resolution = variable.attrs["resolution"]
+
+    variables = {variable_name: variable}
+    index = h3.H3Index.from_variables(variables, options=options)
+
+    assert index._resolution == expected_resolution
+    assert (index._dim,) == variable.dims
+
+    # TODO: how do we check the index, if at all?
+    assert (index._pd_index.dim,) == variable.dims
+    assert np.all(index._pd_index.index.values == variable.data)
+
+
+@pytest.mark.parametrize(["old_variable", "new_variable"], variable_combinations)
+def test_replace(old_variable, new_variable):
+    index = h3.H3Index(
+        cell_ids=old_variable.data,
+        dim=old_variable.dims[0],
+        resolution=old_variable.attrs["resolution"],
+    )
+    new_pandas_index = PandasIndex.from_variables(
+        {"cell_ids": new_variable}, options={}
+    )
+
+    new_index = index._replace(new_pandas_index)
+
+    assert new_index._resolution == index._resolution
+    assert new_index._dim == index._dim
+    assert new_index._pd_index == new_pandas_index
+
+
+@pytest.mark.parametrize(
+    ["cell_ids", "cell_centers"], list(zip(cell_ids, cell_centers))
+)
+def test_cellid2latlon(cell_ids, cell_centers):
+    index = h3.H3Index(cell_ids=cell_ids, dim="cells", resolution=3)
+
+    actual = index._cellid2latlon(cell_ids)
+    expected = cell_centers
+
+    np.testing.assert_allclose(actual, expected)
+
+
+@pytest.mark.parametrize(
+    ["cell_centers", "cell_ids"], list(zip(cell_centers, cell_ids))
+)
+def test_latlon2cellid(cell_centers, cell_ids):
+    index = h3.H3Index(cell_ids=[0], dim="cells", resolution=3)
+
+    actual = index._latlon2cellid(cell_centers[:, 0], cell_centers[:, 1])
```

Comment:
the pattern of creating dummy data that is then not used to me indicates that there's something wrong with the API: we only need resolution and lat / lon values for the computation, so I'd either make these functions (and pass `self._resolution`), or have the subclasses override the methods / properties directly.

For testing, the former would be easier (but makes the implementation a bit clunkier)

Reply:
I'm actually not sure that we should really go down to that level for the unit tests. It may be fine to let this logic be covered in higher-level tests consuming the public API (accessor), especially that we are mostly relying on h3ronpy (and h3o) without implementing much on top of it here.

Reply:
for `h3`, yes, we only forward, whereas for `healpix` there's a bit more work involved. So not sure if this is indeed worth changing, but I do think we need to have *some* tests covering the lower-level API: that way, we can have the tests of the higher-level API be less extensive (like, one or two tests for each implementation).

Let me see if I can come up with something better, otherwise we just have to stick with the dummy data in the tests.